ambry-cloud/src/main/java/com.github.ambry.cloud/CloudBlobStore.java

```diff
@@ -33,40 +34,53 @@
 import java.io.IOException;
 import java.nio.ByteBuffer;
 import java.nio.channels.ReadableByteChannel;
+import java.security.GeneralSecurityException;
 import java.util.EnumSet;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.Set;
 import java.util.concurrent.TimeUnit;
 import java.util.stream.Collectors;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import static com.github.ambry.cloud.CloudBlobMetadata.EncryptionOrigin;
 
 /**
  * The blob store that reflects data in a cloud storage.
  */
 class CloudBlobStore implements Store {
 
-  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private static final Logger logger = LoggerFactory.getLogger(CloudBlobStore.class);
   private final PartitionId partitionId;
   private final CloudDestination cloudDestination;
   private final CloudConfig cloudConfig;
+  private final CloudBlobCryptoAgentFactory cryptoAgentFactory;
+  private final CloudBlobCryptoAgent cryptoAgent;
+  private final VcrMetrics vcrMetrics;
   private final long minTtlMillis;
   private boolean started;
 
   /**
    * Constructor for CloudBlobStore
    * @param partitionId partition associated with BlobStore.
+   * @param cloudDestination the {@link CloudDestination}.
    * @param cloudConfig the {@link CloudConfig} to use.
    * @param cloudDestination the {@link CloudDestination} to use.
+   * @param cryptoAgentFactory the {@link CloudBlobCryptoAgentFactory} to use for encryption.
+   * @param vcrMetrics the {@link VcrMetrics} to use.
    */
-  CloudBlobStore(PartitionId partitionId, CloudConfig cloudConfig, CloudDestination cloudDestination) {
+  CloudBlobStore(PartitionId partitionId, CloudConfig cloudConfig, CloudDestination cloudDestination,
+      CloudBlobCryptoAgentFactory cryptoAgentFactory, VcrMetrics vcrMetrics) {
```

Comment:
Try not to pass in `cryptoAgentFactory` . Let's talk offline.

Reply:
Reasons:
1. CloudBackupManager and CloudBlobStore interfaces looks cleaner.
2. For CloudBlobStore.get(), we need to dynamically load `CloudBlobCryptoAgentFactory` anyway. 

So I think we can pass in `property` and `registry` instead of a `CloudBlobCryptoAgentFactory`.